apf/apfSimplexAngleCalcs.cc

```diff
@@ -0,0 +1,366 @@
+}
+
+static Vector3 computeFaceNormalAtEdge(Mesh* m, /*MeshEntity* tet,*/
+    MeshEntity* face, MeshEntity* edge, Matrix3x3 Q)
+{
+  PCU_ALWAYS_ASSERT(m->getType(face) == Mesh::TRIANGLE);
+  PCU_ALWAYS_ASSERT(m->getType(edge) == Mesh::EDGE);
+
+  MeshEntity* fe[3];
+  m->getDownward(face, 1, fe);
+  int index = -1;
+  for (int i = 0; i < 3; i++)
+    if (fe[i] == edge) {
+      index = i;
+      break;
+    }
+  PCU_ALWAYS_ASSERT(index > -1 && index < 3);
+  Vector3 xi; // corresponding parametric coord in tri of the mid-edge point
+  switch (index) {
+    case 0:
+      xi = Vector3(0.5, 0.0, 0.0);
+      break;
+    case 1:
+      xi = Vector3(0.5, 0.5, 0.0);
+      break;
+    case 2:
+      xi = Vector3(0.0, 0.5, 0.0);
+      break;
+    default:
+      break;
+  }
+  /* Vector3 xi_2 = apf::boundaryToElementXi(m, edge, face, Vector3(0.0, 0.0, 0.0)); */
+  /* std::cout << "\n" << xi << " " <<  xi_2 << " " << (xi_2 - xi).getLength() << "\n" << std::endl; */
+  MeshElement* faceElem = createMeshElement(m, face);
+  Matrix3x3 J;
+  getJacobian(faceElem, xi, J);
+  destroyMeshElement(faceElem);
+
+  /* Vector3 xi_tet = apf::boundaryToElementXi(m, edge, tet, Vector3(0.0, 0.0, 0.0)); // apf::boundaryToElementXi(m, face, tet, xi); */
+  /* MeshElement* tetElem = createMeshElement(m, tet); */
+  /* Matrix3x3 JTet; */
+  /* getJacobian(tetElem, xi_tet, JTet); */
+  /* destroyMeshElement(tetElem); */
+
+  // transform Jacobian into Metric space
+  J = getJacobianInMetric(J, Q);
+  return cross(J[0], J[1]).normalize();
+}
+
+
+static double computeFaceFaceCosAngleInTet(Mesh* m, MeshEntity* tet,
+    MeshEntity* face1, MeshEntity* face2, const Matrix3x3& Q)
+{
+  PCU_ALWAYS_ASSERT(m->getType(tet) == Mesh::TET);
+  PCU_ALWAYS_ASSERT(m->getType(face1) == Mesh::TRIANGLE);
+  PCU_ALWAYS_ASSERT(m->getType(face2) == Mesh::TRIANGLE);
+
+  if (face1 == face2) return 1.0;
+
+  MeshEntity* f1e[3];
+  MeshEntity* f2e[3];
+  MeshEntity* sharedEdge = 0;
+
+  m->getDownward(face1, 1, f1e);
+  m->getDownward(face2, 1, f2e);
+
+  for (int i = 0; i < 3; i++)
+    for (int j = 0; j < 3; j++)
+      if (f1e[i] == f2e[j]) {
+      	sharedEdge = f1e[i];
+      	break;
+      }
+
+  PCU_ALWAYS_ASSERT(sharedEdge);
+  Vector3 f1n = computeFaceNormalAtEdge(m, /*tet,*/ face1, sharedEdge, Q);
+  Vector3 f2n = computeFaceNormalAtEdge(m, /*tet,*/ face2, sharedEdge, Q);
+  // turn f2n so you get the inside angle
+  f2n = Vector3(0.0, 0.0, 0.0) - f2n;
+  return f1n * f2n;
+}
+
+static double computeEdgeFaceCosAngleInTet(Mesh* m, MeshEntity* tet,
+    MeshEntity* edge, MeshEntity* face, const Matrix3x3& Q)
+{
+  PCU_ALWAYS_ASSERT(m->getType(tet) == Mesh::TET);
+  PCU_ALWAYS_ASSERT(m->getType(edge) == Mesh::EDGE);
+  PCU_ALWAYS_ASSERT(m->getType(face) == Mesh::TRIANGLE);
+
+  MeshEntity* fe[3];
+  m->getDownward(face, 1, fe);
+  if (findIn(fe, 3, edge) > -1)
+    return 1.0;
+
+
+  // find the share vertex
+  MeshEntity* ev[2];
+  MeshEntity* fv[3];
+  MeshEntity* sharedVert = 0;
```

Comment:
The trick here is that the MeshEntity* is actually the id of MDS entity.  But, I was wrong and MDS starts ids at 1.  If they did start at 0 then the assert on line 292 could trip on a valid case.